records/2023-06-27.scalability.increasing.send.rates.md

```diff
@@ -0,0 +1,107 @@
+- Prepares the email
+- Get the template
+- Fill it out
+- Get any files if they’re attached
+- Calls `send_email()` with the html and any attachments
+
+[`send_email()`](https://github.com/cds-snc/notification-api/blob/main/app/clients/email/aws_ses.py#L33)
+- Creates a MIME email with the html and attachments
+- Calls boto3 `send_raw_email()` with the MIME email
+
+So there are basically 3 areas that can take time
+- `send_email_to_provider()` code before the `send_email()` call
+- `send_email()` code before the boto call
+- the boto call
+
+We have timing metrics for:
+- `deliver_email()` (which should be the same as `send_email_to_provider()`): celery_tasks_deliver_email
+- `send_email()`: celery_clients_ses_send_email
+- `send_raw_email()`:  celery_clients_ses_request-time
+
+
+We uploaded a csv of 50K emails (with one variable and no file attachments) to staging (send to `success@simulator.amazonses.com`) on July 12, 2:15 pm EST and monitored the [scaling dashboard](https://ca-central-1.console.aws.amazon.com/cloudwatch/home?region=ca-central-1#dashboards/dashboard/ScalingTemp?start=2023-07-12T18%3A10%3A00Z&end=2023-07-12T19%3A30%3A00Z&autoRefresh=false).
+
+
+![bulk-tasks activity](./diagrams/2023-06-27.scalability.increasing.send.rates/bulk_tasks-queue.png "bulk-task queue")
+
+This graph shows the deliver_email tasks are added to the bulk-tasks queue at a rate of about 7500 / minute. They are being processed at a rate of about 750 / minute.
+
+
+![deliver_email time](./diagrams/2023-06-27.scalability.increasing.send.rates/deliver_email-average-time.png "deliver_email average time")
+
+This graph shows that each deliver_email task took approximately 0.3 seconds during the load test.
+
+
+![deliver_email time breakdown](./diagrams/2023-06-27.scalability.increasing.send.rates/deliver_email-breakdown.png "deliver_email breakdown")
+
+From this graph we see that during the deliver_email task:
+- 65% of the time is in deliver_email code before calling send_email
+- 33% of the time is spent doing the boto call
+- 2% of the time is spent doing send_email code before the boto call
+
+
+## Options
+
+There are two basic areas that can be changed to increase our send rates:
+- Notify code
+- Notify infrastructure
+
+### Notify code
+
+We can try to speed up:
+- the `deliver_email` code before the `send_email` call
+- the boto call (can it be batched somehow?)
+
+Note that without speeding up the boto call our improvements will be limited.
+
+### Notify infrastructure
+
+Our approach is already parallelized (an asyncronous task is created for each notification to be delivered). Simply running more tasks at once (ie more workers) would speed things up. This could of course create another bottleneck (we've had issues with the database in the past, for example)
+
+One approach we might want to investigate is creating more celery workers and dedicating them to `deliver_email` and / or `deliver_sms`.
```

Comment:
this feels like a reasonable approach.. is there a diminishing return graph on this?

Reply:
it should be pretty much a linear improvement until we start bottlenecking somewhere else (ie database, network). definitely something we want to examine.